### ISSUE TITLE
[chopper_built_value] @Query does not use serializers to serialize query params

### ISSUE THREAD

**Issue by danielgomezrico** [lejard-h/chopper#332]
# Current
I use BuiltValue with EnumClass to have nice enums, but Im trying to use an enum as a `@Query` param and it uses the `toString` value in the request instead of using the serializer value.

I have an Enum that is declared like:
```dart
class TurnVisitType extends EnumClass {
  const TurnVisitType._(String name) : super(name);

  static Serializer<TurnVisitType> get serializer => _$turnVisitTypeSerializer;
...

  @BuiltValueEnumConst(wireName: 'face_to_face')
  static const TurnVisitType faceToFace = _$faceToFace;

  static const TurnVisitType virtual = _$virtual;
```

The generated serializer:
```dart
class _$TurnVisitTypeSerializer implements PrimitiveSerializer<TurnVisitType> {
  static const Map<String, Object> _toWire = const <String, Object>{
    'faceToFace': 'face_to_face',
  };
  static const Map<Object, String> _fromWire = const <Object, String>{
    'face_to_face': 'faceToFace',
  };

  @override
  final Iterable<Type> types = const <Type>[TurnVisitType];
  @override
  final String wireName = 'TurnVisitType';

  @override
  Object serialize(Serializers serializers, TurnVisitType object,
          {FullType specifiedType = FullType.unspecified}) {
    return _toWire[object.name] ?? object.name;
  }
```

And my service like this:
```dart
  @Get(path: 'api/v1/available_turn_schedules/{id}')
  Future<Response<List<ScheduleResponse>>> getSchedules(
      @Path('id') int realEstateId,
      @Query('visity_type') TurnVisitType visitType);
```

But when I see how the request is done, it sends `faceToFace` (`visitType.toString()` value) instead of the serializer `TurnVisitType.serializer` and the `wiredValue` and send `face_to_face`.

# Some places on chopper

It does use `toString` on every query param, and it does not try to use the serializes.

Here:
https://github.com/lejard-h/chopper/blob/f558bc7aabc36e819611cc7001d8f37f59dae309/chopper/lib/src/request.dart#L126

Here:
https://github.com/lejard-h/chopper/blob/82b951fe278fcd783de5285b4ef2b7cbbede61c9/chopper/lib/src/utils.dart#L76-L82

# Expected
The url is built using the serializer if the type have it declared, and

**Comment by danielgomezrico** [comment]
👀

**Comment by JEuler** [comment]
Sorry, just dropped out of life a bit. Will have a look later.

**Comment by techouse** [comment]
@danielgomezrico is this still an issue?

**Comment by danielgomezrico** [comment]
@techouse yeah

**Comment by techouse** [comment]
Right, so this is due to [this line of code](https://github.com/lejard-h/chopper/blob/master/chopper/lib/src/utils.dart#L132)

```dart
String _normalizeValue(value) => Uri.encodeComponent(value?.toString() ?? '');
```

Do you have any suggestions on how to make this work with BuiltValue? I'm not (yet) familiar with BuiltValue as I've always used json_serializable.

**Comment by techouse** [comment]
Getting to the bottom of this slowly. From some initial tests I've written, the support for these `built_value` serializers only extends to encoding and decoding the payload, not the query params.

I'll keep digging and see what can and can not be done.

**Comment by radzish** [comment]
I am missing this too. What is even worse is that I can not write my own converter to handle query params as I need. Request converter will be only applied when body is not empty.

**Comment by techouse** [comment]
> I am missing this too. What is even worse is that I can not write my own converter to handle query params as I need.

You can always submit a PR 🤗

**Comment by dustin-graham** [comment]
Ran into this problem with regular enums. One way to work around this is to write custom toString() method:

```dart
@override
String toString() {
  return this.name;
}
```

**Comment by techouse** [comment]
@dustin-graham yep, that's how I always work with it.

I'll take another look at this over the Xmas holidays unless someone beats me to it 🤞

**Comment by techouse** [comment]
Not sure if this resolves this particular custom serialization issue, but last week I replaced the old query serialization with my own port of [qs](https://github.com/techouse/qs) in https://github.com/lejard-h/chopper/pull/592.

CC / @dustin-graham @radzish @danielgomezrico

**Comment by techouse** [comment]
Sooooooooo .... I finally had the chance (and remembered) to look at this again and came up with a ~draft~ solution in PR #712. 😇

Would you guys mind testing the PR branch directly and giving some feedback?

CC / @danielgomezrico @radzish @dustin-graham @JEuler
